### DISCUSSION TITLE
Dataframe type in typst ?

### DISCUSSION THREAD

**Discussion by DrLoiseau** [typst/typst#3188] [Ideas]
Hi,

Are there plans to add dataframes in typst ? I was thinking it would be great to have a dataframe type and methods to work on it, in the spirit of dplyr in R https://dplyr.tidyverse.org/ : you can easily add new rows or columns, compute new columns based on calculations on existing columns, group and summarise data with aggregation function, filter, etc. and do all of this sequentially with a pipe operator, so the syntax and the different steps to produce the dataframe appear very clearly.
Since typst arrays already have **many** cool features like filter, sum, fold and zip, it looks like it would be possible to add dataframe methods to help work on arrays of arrays.

I've seen the package TaDa https://github.com/typst/packages/tree/main/packages/preview/tada/0.1.0 that brings some of these features, but it is mostly oriented towards table formatting, not dataframe manipulations. I think it is better to separate dataframe manipulation and table formatting.

Would this kind of features fit into typst, or is it better to use external code in notebooks, like it was advised in this post [#2270](https://github.com/typst/typst/discussions/2270) for pandas  ?

**Comment by laurmaedje** [comment]
This is unlikely to become built-in. Maybe someone will make a more generalized dataframe package at some point.

**Comment by felipeacsi** [comment]
> I've seen the package TaDa https://github.com/typst/packages/tree/main/packages/preview/tada/0.1.0 that brings some of these features, but it is mostly oriented towards table formatting, not dataframe manipulations. I think it is better to separate dataframe manipulation and table formatting.

I think is more than that... Have you seen [this answer on Discord](https://discord.com/channels/1054443721975922748/1190824934700765255/1192720948156170320) from the developer of `tada`? He said:

> Once Typst allows custom types (see ⁠Type and content rework), the API should get *significantly* better

Maybe that entire question on Discord may be of help.

I also have "frustrations" with `tada`, but it is a very good initial release.
